File: backend/app/schemas/digest.py
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field

from app.enums import Platform

DigestPrimaryState = Literal[
    "urgent",
    "needs_igor",
    "needs_reply",
    "waiting",
    "new_activity",
    "resolved",
    "informational",
]


class DigestPeriod(BaseModel):
    label: str
    start: datetime
    end: datetime


class DigestCounts(BaseModel):
    messages: int = 0
    incoming: int = 0
    outgoing: int = 0
    active_chats: int = 0
    needs_reply: int = 0
    needs_igor: int = 0
    urgent: int = 0
    waiting: int = 0
    resolved: int = 0
    igor_participated: int = 0
    waiting_for_us: int = 0
    waiting_for_them: int = 0


class DigestItem(BaseModel):
    chat_id: int
    platform: Platform
    chat_name: str
    status: str
    target_message_id: int | None = None
    latest_message_at: datetime | None = None
    primary_state: DigestPrimaryState
    needs_reply: bool = False
    needs_igor: bool = False
    urgent: bool = False
    waiting: bool = False
    resolved: bool = False
    already_answered: bool = False
    high_stakes: bool = False
    analysis_available: bool = False
    analysis_fresh: bool = False
    summary_ru: str = ""
    next_action_ru: str = ""
    snippet: str = ""
    snippet_translated: str | None = None
    source_message_count: int = 0
    igor_participated: bool = False
    period_outgoing_count: int = 0


class DigestAIEntry(BaseModel):
    chat_id: int
    message_id: int | None = None
    source_message_ids: list[int] = Field(default_factory=list)
    title_ru: str = ""
    summary_ru: str = ""
    next_action_ru: str = ""
    action_ru: str = ""


class DigestAIAction(DigestAIEntry):
    person_or_chat_ru: str = ""
    result_ru: str = ""
    confidence: Literal["explicit", "strong", "uncertain"] = "uncertain"


class DigestAIInteraction(DigestAIEntry):
    person_or_chat_ru: str = ""
    platform: str = ""
    topic_ru: str = ""
    what_happened_ru: str = ""
    igor_last_action_ru: str = ""
    current_state_ru: str = ""


class DigestAIFact(BaseModel):
    chat_id: int
    message_id: int | None = None
    source_message_ids: list[int] = Field(default_factory=list)
    fact_ru: str = ""


class DigestAIPeriodStats(BaseModel):
    active_chats: int = 0
    messages: int = 0
    igor_participated_chats: int = 0
    waiting_for_us: int = 0
    waiting_for_them: int = 0


class DigestAIOutput(BaseModel):
    model_config = {"extra": "ignore"}

    title_ru: str = ""
    executive_summary_ru: str = ""
    period_stats: DigestAIPeriodStats = Field(default_factory=DigestAIPeriodStats)
    main_events: list[DigestAIEntry] = Field(default_factory=list)
    igor_actions: list[DigestAIAction] = Field(default_factory=list)
    interactions: list[DigestAIInteraction] = Field(default_factory=list)
    needs_action: list[DigestAIEntry] = Field(default_factory=list)
    waiting_for_others: list[DigestAIEntry] = Field(default_factory=list)
    completed_or_answered: list[DigestAIEntry] = Field(default_factory=list)
    results_and_numbers: list[DigestAIFact] = Field(default_factory=list)
    blockers_and_risks: list[DigestAIEntry] = Field(default_factory=list)
    next_steps: list[DigestAIEntry] = Field(default_factory=list)


class DigestAICacheInfo(BaseModel):
    available: bool = False
    stale: bool = False
    created_at: datetime | None = None
    result: DigestAIOutput | None = None
    model: str | None = None


class DigestResponse(BaseModel):
    period: DigestPeriod
    counts: DigestCounts
    items: list[DigestItem]
    source_hash: str
    ai: DigestAICacheInfo = Field(default_factory=DigestAICacheInfo)


class DigestAIRequest(BaseModel):
    period: str | None = "24h"
    start: datetime | None = None
    end: datetime | None = None
    platform: Platform | None = None
    force: bool = False
    model: str | None = None


class DigestAIResponse(BaseModel):
    period: DigestPeriod
    source_hash: str
    cached: bool
    stale: bool = False
    result: DigestAIOutput
    provider: str | None = None
    model: str | None = None


class DigestQAHistoryTurn(BaseModel):
    role: Literal["user", "assistant"]
    content: str = ""


class DigestQARequest(BaseModel):
    period: str | None = "24h"
    start: datetime | None = None
    end: datetime | None = None
    model: str | None = None
    question: str
    history: list[DigestQAHistoryTurn] = Field(default_factory=list)


class DigestQASource(BaseModel):
    chat_id: int
    message_id: int
    platform: Platform
    chat_name: str
    timestamp: datetime | None = None


class DigestQAContextStats(BaseModel):
    chats: int = 0
    messages: int = 0


class DigestContextSnapshot(BaseModel):
    """Safe selected-source refs. IDs and hashes only — never full duplicated bodies."""

    message_ids: list[int] = Field(default_factory=list)
    chat_ids: list[int] = Field(default_factory=list)
    aliases: dict[str, int] = Field(default_factory=dict)
    text_hashes: dict[str, str] = Field(default_factory=dict)
    question: str | None = None
    period_label: str | None = None
    period_start: str | None = None
    period_end: str | None = None
    truncated: bool = False


class DigestQAModelOutput(BaseModel):
    model_config = {"extra": "ignore"}

    answer_ru: str = ""
    source_refs: list[str] = Field(default_factory=list)
    uncertainty_ru: str | None = None
    suggested_questions_ru: list[str] = Field(default_factory=list)


class DigestQAResponse(BaseModel):
    answer_ru: str
    sources: list[DigestQASource] = Field(default_factory=list)
    model: str
    context_stats: DigestQAContextStats = Field(default_factory=DigestQAContextStats)
    uncertainty_ru: str | None = None
    suggested_questions_ru: list[str] = Field(default_factory=list)
    context_snapshot: DigestContextSnapshot | None = None


class DigestQAExportRequest(BaseModel):
    format: Literal["md", "json"] = "md"
    period: str | None = "24h"
    start: datetime | None = None
    end: datetime | None = None
    model: str | None = None
    question: str = ""
    history: list[DigestQAHistoryTurn] = Field(default_factory=list)
    snapshot: DigestContextSnapshot | None = None


class DigestAICandidate(BaseModel):
    chat_id: int
    platform: str
    chat_name: str
    primary_state: str
    needs_reply: bool
    needs_igor: bool
    urgent: bool
    waiting: bool
    already_answered: bool
    high_stakes: bool
    analysis_fresh: bool
    latest_message_at: str | None = None
    snippet: str
    summary_ru: str
    next_action_ru: str
    target_message_id: int | None = None
    source_message_count: int = 0
    igor_participated: bool = False
    period_outgoing_count: int = 0
    messages: list[dict[str, Any]] = Field(default_factory=list)

File: frontend/src/types/digest.ts
import type { Platform } from "./inbox";

export type DigestPeriodLabel = "1h" | "3h" | "6h" | "12h" | "24h" | "3d" | "7d" | "custom";

export type DigestPrimaryState =
  | "urgent"
  | "needs_igor"
  | "needs_reply"
  | "waiting"
  | "new_activity"
  | "resolved"
  | "informational";

export type DigestPeriod = {
  label: string;
  start: string;
  end: string;
};

export type DigestCounts = {
  messages: number;
  incoming: number;
  outgoing: number;
  active_chats: number;
  needs_reply: number;
  needs_igor: number;
  urgent: number;
  waiting: number;
  resolved: number;
  igor_participated: number;
  waiting_for_us: number;
  waiting_for_them: number;
};

export type DigestItem = {
  chat_id: number;
  platform: Platform;
  chat_name: string;
  status: string;
  target_message_id: number | null;
  latest_message_at: string | null;
  primary_state: DigestPrimaryState;
  needs_reply: boolean;
  needs_igor: boolean;
  urgent: boolean;
  waiting: boolean;
  resolved: boolean;
  already_answered: boolean;
  high_stakes: boolean;
  analysis_available: boolean;
  analysis_fresh: boolean;
  summary_ru: string;
  next_action_ru: string;
  snippet: string;
  snippet_translated: string | null;
  source_message_count: number;
  igor_participated: boolean;
  period_outgoing_count: number;
};

export type DigestAIEntry = {
  chat_id: number;
  message_id: number | null;
  source_message_ids?: number[];
  title_ru: string;
  summary_ru: string;
  next_action_ru: string;
  action_ru: string;
};

export type DigestAIAction = DigestAIEntry & {
  person_or_chat_ru: string;
  result_ru: string;
  confidence: "explicit" | "strong" | "uncertain";
};

export type DigestAIInteraction = DigestAIEntry & {
  person_or_chat_ru: string;
  platform: string;
  topic_ru: string;
  what_happened_ru: string;
  igor_last_action_ru: string;
  current_state_ru: string;
};

export type DigestAIFact = {
  chat_id: number;
  message_id: number | null;
  source_message_ids?: number[];
  fact_ru: string;
};

export type DigestAIPeriodStats = {
  active_chats: number;
  messages: number;
  igor_participated_chats: number;
  waiting_for_us: number;
  waiting_for_them: number;
};

export type DigestAIOutput = {
  title_ru: string;
  executive_summary_ru: string;
  period_stats: DigestAIPeriodStats;
  main_events: DigestAIEntry[];
  igor_actions: DigestAIAction[];
  interactions: DigestAIInteraction[];
  needs_action: DigestAIEntry[];
  waiting_for_others: DigestAIEntry[];
  completed_or_answered: DigestAIEntry[];
  results_and_numbers: DigestAIFact[];
  blockers_and_risks: DigestAIEntry[];
  next_steps: DigestAIEntry[];
};

export type DigestAICacheInfo = {
  available: boolean;
  stale: boolean;
  created_at: string | null;
  result: DigestAIOutput | null;
  model: string | null;
};

export type DigestResponse = {
  period: DigestPeriod;
  counts: DigestCounts;
  items: DigestItem[];
  source_hash: string;
  ai: DigestAICacheInfo;
};

export type DigestAIResponse = {
  period: DigestPeriod;
  source_hash: string;
  cached: boolean;
  stale: boolean;
  result: DigestAIOutput;
  provider: string | null;
  model: string | null;
};

export type DigestStateFilter = "all" | "needs_reply" | "needs_igor" | "urgent" | "waiting" | "resolved";
export type DigestPlatformFilter = "all" | Platform;

export type AIModelInfo = {
  id: string;
  label: string;
  description: string;
  cost_level: 1 | 2 | 3 | number;
  recommended_for: string;
};

export type AIModelsResponse = {
  models: AIModelInfo[];
  review_default: string;
  qa_default: string;
};

export type DigestQAHistoryTurn = {
  role: "user" | "assistant";
  content: string;
};

export type DigestQASource = {
  chat_id: number;
  message_id: number;
  platform: Platform;
  chat_name: string;
  timestamp: string | null;
};

export type DigestQAResponse = {
  answer_ru: string;
  sources: DigestQASource[];
  model: string;
  context_stats: { chats: number; messages: number };
  uncertainty_ru: string | null;
  suggested_questions_ru: string[];
  context_snapshot?: DigestContextSnapshot | null;
};

export type DigestContextSnapshot = {
  message_ids: number[];
  chat_ids: number[];
  aliases: Record<string, number>;
  text_hashes: Record<string, string>;
  question?: string | null;
  period_label?: string | null;
  period_start?: string | null;
  period_end?: string | null;
  truncated?: boolean;
};

export type ExportFormat = "md" | "json";


